### DISCUSSION TITLE
Steam DepotDownloader account/password and other parameters

### DISCUSSION THREAD

**Discussion by Linx-ESP** [pufferpanel/pufferpanel#1239] [Q&A] [answered]
How could I use some of the following [parameters](https://github.com/SteamRE/DepotDownloader#parameters) for Steam installs.
In my case it would be for username/password but it could be helfpul for people that want to use other branches.

To further explain my situtation:

> Starting process: /var/lib/pufferpanel/binaries/dotnet-runtime/dotnet /var/lib/pufferpanel/binaries/depotdownloader/DepotDownloader.dll -app 302550 -dir /var/lib/pufferpanel/servers/65133ac5
> =No username given. Using anonymous account with dedicated server subscription.
> Connecting to Steam3... Done!
> Logging anonymously into Steam3... Done!
> Using Steam3 suggested CellID: 40
> Got AppInfo for 302550
> App 302550 (Assetto Corsa Dedicated Server) is not available from this account.
> Disconnected from Steam

Same output using DepotInstaller from a PC without login. 
Using `-username [username] -remember-password` works and keeps the login even with Steam 2FA.

**Comment by LordRalex** [comment, marked as answer]
If you set "username" and "password" in the operation (like appId), then it'd pass those to the DepotDownloader.

If username is set, then it automatically adds -remember-password to the call.

**Reply by Linx-ESP** [reply]
Hopefully I'm not missing something obvious.

Seems to pass the command properly (output below) with both one variable and one "hardcoded" and two variables for the install command and also putting them directly in the steamgamedl "App Id" like the image below. All with the same error.
![image](https://user-images.githubusercontent.com/51830853/223510308-ddb08711-4bd1-4582-9c9c-185f9c36f7a0.png)

Using pufferpanel via docker on ubuntu server, which might matter since the error mentions `C:\Projects\DepotDownloader\DepotDownloader\Program.cs:line 17` as a Windows path (don't know if it is normal because dotnet).

Also tested with AppID 222840 (L4D dedicated server) because doesn't need credentials. Works without username as anonymus, same error when using -username myusername.
Also tested with -username (empty), same error.
Also tested with password, same error.
Works as intended using DepotDownloader directly on windows and arch linux.
No issues that seem relevant on DepotDownloader's github, at least that I saw in a quick search.

```
  "name": "assetto-corsa",
  "display": "Assetto Corsa",
  "type": "steam",
  "install": [
    {
      "appId": "${appid} -username ${steam-username}",
      "type": "steamgamedl"
    }
  ],
  "run": {
    "stop": "",
    "command": "./acserver",
    "workingDirectory": "",
    "pre": [],
    "post": [],
    "environmentVars": {}
  },
  "data": {
    "appid": {
      "type": "string",
      "display": "App ID",
      "required": true,
      "value": "302550"
    },
    "steam-username": {
      "type": "string",
      "display": "Steam username",
      "required": true,
      "value": "username"
    }
  },
  "supportedEnvironments": [
    {
      "type": "standard"
    }
  ],
  "requirements": {}
}
```

>  Starting process: /var/lib/pufferpanel/binaries/dotnet-runtime/dotnet /var/lib/pufferpanel/binaries/depotdownloader/DepotDownloader.dll -app 302550 -username [REDACTED] -dir /var/lib/pufferpanel/servers/bc611224
> Unhandled exception. System.ArgumentException: 302550 -username [REDACTED] is not a valid value for UInt32. (Parameter 'value')
>  ---> System.FormatException: Input string was not in a correct format.
>    at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type)
>    at System.UInt32.Parse(String s, NumberStyles style, IFormatProvider provider)
>    at System.ComponentModel.UInt32Converter.FromString(String value, NumberFormatInfo formatInfo)
>    at System.ComponentModel.BaseNumberConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value)
>    --- End of inner exception stack trace ---
>    at System.ComponentModel.BaseNumberConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value)
>    at System.ComponentModel.TypeConverter.ConvertFromString(String text)
>    at DepotDownloader.Program.GetParameter[T](String[] args, String param, T defaultValue) in C:\Projects\DepotDownloader\DepotDownloader\Program.cs:line 328
>    at DepotDownloader.Program.MainAsync(String[] args) in C:\Projects\DepotDownloader\DepotDownloader\Program.cs:line 101
>    at DepotDownloader.Program.Main(String[] args) in C:\Projects\DepotDownloader\DepotDownloader\Program.cs:line 17

**Reply by nepcore** [reply]
I _think_ what @LordRalex meant was that you edit the json, currently you have `{"appId": "${appid}", "type": "steamgamedl"}`, so make that `{"appId": "${appid}", "username": "${steam-username}", "type": "steamgamedl"}` to add a username to the download command

**Reply by Linx-ESP** [reply]
Yep, that worked.
Now I have the issue of 2FA since it cannot be passed the same way that username and password are, [at least I don't see how](https://github.com/SteamRE/DepotDownloader#parameters).
Writing them in console doesn't work, neither `echo "code"`. Should be possible via the console any other way?

**Reply by LordRalex** [reply]
I believed the console should still function during such a process, but I'd have to re-check.

It is the same code that running a server is, it's not like it's a "special" prompt.
